<launch>

  <arg name="debug" default="0"/>
  <arg name="mode" default="default"/>
  <arg name="namespace" default="floatsam"/>
  <arg name="odom_topic" default="/$(arg namespace)/dr/odom_raw" />
  <arg name="gps_odom_topic" default="/$(arg namespace)/dr/gps_odom"/>
  
  <group ns="$(arg namespace)">

      <!-- <param name="robot_description" 
          command="$(find xacro)/xacro.py '$(find sam_description)/robots/sam_auv_$(arg mode).urdf.xacro' debug:=$(arg debug) namespace:=sam_auv" />

      <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" respawn="true" output="screen">
        <param name="use_gui" value="false"/>
        <rosparam param="source_list">["/sam/command_states"]</rosparam>
      </node>

      <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="true" output="screen">
        <param name="robot_description" value="/$(arg namespace)/robot_description" />
      </node> -->
      
      <group ns="dr">

        <node pkg="tf_lat_lon" type="tf_lat_lon_node" name="tf_lat_lon_node" output="screen">
          <param name="frame" value="$(arg namespace)/base_link"/>
          <param name="utm_zone" value="$(arg utm_zone)"/>
          <param name="utm_band" value="$(arg utm_band)"/>
          <remap from="tf_lat_lon" to="lat_lon"/>
        </node>

        <node pkg="floatsam_dead_reckoning" name="gps_pose_pub" type="usv_gps_node.py" respawn="true" output="screen">
          <param name="map_frame" value="map" />
          <param name="utm_frame" value="utm" />
          <param name="base_frame" value="$(arg namespace)/base_link" />
          <param name="gps_main_frame" value="$(arg namespace)/gps_link" />
          <param name="rtk_prt_frame" value="$(arg namespace)/rtk_prt_link" />
          <param name="rtk_stb_frame" value="$(arg namespace)/rtk_stb_link" />
          <!-- Output topics -->
          <param name="rtk_odom_topic" value="/$(arg namespace)/dr/gps_aux_odom" />
          <param name="aux_gps_odom_topic" value="$(arg gps_odom_topic)" />
          <param name="rtk_prt_odom_top" value="/$(arg namespace)/dr/rtk_prt_odom" />
          <param name="rtk_stb_odom_top" value="/$(arg namespace)/dr/rtk_stb_odom" />
          <!-- Input topics -->
          <param name="rtk_prt_topic" value="/$(arg namespace)/core/rtk_prt" />
          <param name="rtk_stb_topic" value="/$(arg namespace)/core/rtk_stb" />
          <param name="aux_gps_topic" value="/$(arg namespace)/core/gps" />
          
        </node>
        
        <node pkg="floatsam_dead_reckoning" name="dr_node" type="usv_dr_node.py" output="screen">
          <param name="odom_topic" value="$(arg odom_topic)" />
          <param name="stim_imu" value="/$(arg namespace)/core/imu" />
          <!--<param name="sbg_topic" value="/$(arg namespace)/sbg/ekf_quat" />-->
          <param name="sbg_topic" value="/$(arg namespace)/core/sbg_imu" />
          <param name="thrust1_fb" value="/$(arg namespace)/core/thruster1_fb" />
          <param name="thrust2_fb" value="/$(arg namespace)/core/thruster2_fb" />
          <param name="gps_odom_topic" value="$(arg gps_odom_topic)" />
          <param name="odom_frame" value="$(arg namespace)/odom" />
          <param name="utm_frame" value="utm" />
          <param name="base_frame" value="$(arg namespace)/base_link" />
          <param name="base_frame_2d" value="$(arg namespace)/base_link_2d" />
          <param name="map_frame" value="map" />
          <!-- <param name="dvl_frame" value="$(arg namespace)/dvl_link" /> -->
          <!-- <param name="dvl_period" value="0.1" /> -->
          <param name="dr_period" value="0.01" />
          <param name="KT_thrusters" value="0.001" />
        </node>
      
        <include file="$(find auv_particle_filter)/launch/auv_pf.launch">
            <arg name="namespace" value="$(arg namespace)"/>
            <arg name="map_frame" value="map"/>
            <arg name="utm_frame" value="utm"/>
            <arg name="odom_frame" value="$(arg namespace)/odom"/>
            <arg name="base_frame" value="$(arg namespace)/base_link" />
            <arg name="odom_topic" value="$(arg odom_topic)" />
            <arg name="gps_odom_topic" value="$(arg gps_odom_topic)" />
            <arg name="base_frame_2d" value="$(arg namespace)/base_link_2d" />
            <arg name="particle_count"  value="50"/>
        </include>

        <!-- <node pkg="auv_particle_filter" name="visual_tools" type="visual_tools.py" output="screen">
          <param name="dvl_dr_topic" value="/$(arg namespace)/dr/dvl_dr" />
          <param name="odom_corrected_topic" value="/$(arg namespace)/dr/odom_corrected"/>
          <param name="gps_odom_topic" value="/$(arg namespace)/dr/gps_odom_sam" />
        </node> -->

    	  <node name="odom_listener" pkg="tf_convenience_topics" type="odom_listener" output="screen" >
	    	  <param name="odom_topic" value="/$(arg namespace)/dr/odom"/>
	    	  <param name="topic_from_dvl" value="$/(arg namespace)/core/dvl"/>
    	  </node>

        <!-- Temp: provides heading for nodered -->
        <node pkg="sam_dead_reckoning" name="yaw_2_heading" type="yaw_2_heading.py" output="screen"/>

        <!-- Rviz vis of latest WP from nodered -->
        <node pkg="sam_dead_reckoning" name="wp_vis_node" type="wp_vis_node.py" output="screen"/>


      </group>
    </group>

</launch>
